app/uk/gov/hmrc/transitmovementsrouter/services/ObjectStoreService.scala

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.transitmovementsrouter.services
+
+import cats.data.EitherT
+import cats.implicits._
+import com.google.inject.ImplementedBy
+import play.api.Logging
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.objectstore.client.ObjectSummaryWithMd5
+import uk.gov.hmrc.objectstore.client.Path
+import uk.gov.hmrc.objectstore.client.play.PlayObjectStoreClientEither
+import uk.gov.hmrc.transitmovementsrouter.models.MessageId
+import uk.gov.hmrc.transitmovementsrouter.models.MovementId
+import uk.gov.hmrc.transitmovementsrouter.models.errors.ObjectStoreError
+import uk.gov.hmrc.transitmovementsrouter.models.responses.UpscanResponse.DownloadUrl
+
+import java.net.URL
+import java.time.Clock
+import java.time.OffsetDateTime
+import java.time.ZoneOffset
+import java.time.format.DateTimeFormatter
+import javax.inject.Inject
+import javax.inject.Singleton
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import java.net.URL
+import java.time.Clock
+import java.time.OffsetDateTime
+import java.time.ZoneOffset
+import java.time.format.DateTimeFormatter
+import javax.inject.Inject
+import javax.inject.Singleton
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.Try
+import scala.util.control.NonFatal
+
+@ImplementedBy(classOf[ObjectStoreServiceImpl])
+trait ObjectStoreService {
+
+  def addMessage(upscanUrl: DownloadUrl, movementId: MovementId, messageId: MessageId)(implicit
+    hc: HeaderCarrier,
+    ec: ExecutionContext
+  ): EitherT[Future, ObjectStoreError, ObjectSummaryWithMd5]
+
+}
+
+@Singleton
+class ObjectStoreServiceImpl @Inject() (clock: Clock, client: PlayObjectStoreClientEither) extends ObjectStoreService with Logging {
+
+  private val dateTimeFormatter = DateTimeFormatter.ofPattern("yyyyMMdd-HHmmss").withZone(ZoneOffset.UTC)
```

Comment:
I noticed few points with this class:

- will there be any impact if we use ``` Clock.systemUTC().instant ``` directly. Instead of configuring it with Module.scala. I feel that systemUTC is a static method so no need to Inject it with this class and can be used directly.
-  ``` upscanUrl: DownloadUrl ``` DownloadUrl we receive from Upscan with UpscanSuccess response so url is also going to be meaningful. so is it worth to put ``` new URL(upscanUrl.value) ``` inside ``` Future.fromTry ```?.
I feel that we can directly use new URL(upscanUrl.value) without Try to make it more simpler.
-  As we are handling only UnexpectedError so is it not worth to use  ``` PlayObjectStoreClient ``` instead of ``` PlayObjectStoreClientEither ''' to make this class more simple and clean.
-  with below 
``` osClient.uploadFromUrl(
  from            = new URL("https://fus-outbound.s3.eu-west-2.amazonaws.com/81fb03f5-195d-422a-91ab-460939045846"),
  to              = Path.File(Path.Directory("my-folder"), "sample.pdf"),
  retentionPeriod = RetentionPeriod.OneWeek, // defaults to 'object-store.default-retention-period' configuration
  contentType     = Some("text/csv"), // defaults to None which results in the contentType 'application/octet-stream'
  contentMd5      = Some(Md5Hash("a3c2f1e38701bd2c7b54ebd7b1cd0dbc")), // defaults to None
  owner           = "my-service" // defaults to 'appName' configuration
) // returns Future[ObjectSummaryWithMd5] ``` I observed the owner is basically by default is going to be 'appName' so in our case with this service it is going to be 'transit-movements-router' but need this to be overridden with 'common-transit-convention-traders'.  so we need to pass explicitly 'common-transit-convention-traders' with fifth parameter 'owner'.

Below code snippet looks more simpler and meaningful to me.

``` @Singleton
class ObjectStoreServiceImpl @Inject() (client: PlayObjectStoreClient) extends ObjectStoreService{

  private val dateTimeFormatter = DateTimeFormatter.ofPattern("yyyyMMdd-HHmmss").withZone(ZoneOffset.UTC)

  override def uploadToObjectStore(upscanUrl: DownloadUrl, movementId: MovementId, messageId: MessageId)(implicit
    hc: HeaderCarrier,
    ec: ExecutionContext
  ): EitherT[Future, ObjectStoreError, ObjectSummaryWithMd5] =
    EitherT {
      val formattedDateTime = dateTimeFormatter.format(OffsetDateTime.ofInstant(Clock.systemUTC().instant, ZoneOffset.UTC))
      client
        .uploadFromUrl(
          from = new URL(upscanUrl.value),
          to = Path.Directory(s"movements/${movementId.value}").file(s"${movementId.value}-${messageId.value}-$formattedDateTime.xml"),
          retentionPeriod = RetentionPeriod.SevenYears,
          None,
          None,
          owner = "common-transit-convention-traders"
        )
        .map {
          load => Right(load)
        }
        .recover {
          case NonFatal(thr) => Left(ObjectStoreError.UnexpectedError(thr = Some(thr)))
        }
    }
} ```

Reply:
> will there be any impact if we use Clock.systemUTC().instant directly. Instead of configuring it with Module.scala. I feel that systemUTC is a static method so no need to Inject it with this class and can be used directly.

We configure it so we can provide a test clock for testing. Such a clock can be given a static, known, value for example. So yes, we should inject it to give ourselves that flexibility.

> I feel that we can directly use new URL(upscanUrl.value) without Try to make it more simpler.

It really depends on the context, but the general rule is to never trust the client. If this isn't a URL, it's not our fault, but theirs, so we should be returning a 400 to them. Without catching it, it'll be a 500.

But it all depends on the flow. If we're mapping futures (like I think we are here), then it makes the code flow better and may even simplify it.

> As we are handling only UnexpectedError so is it not worth to use PlayObjectStoreClient instead of PlayObjectStoreClientEither to make this class more simple and clean.

That does make sense, but equally, is there anything that might come out of the client that we do need to catch? I see why we'd use an `Either` here because it is more idomatic Scala.

Could go either way on that.

> Below code snippet looks more simpler and meaningful to me.

I'd rather we didn't just blow up on `new URL`. Equally, that `uploadFromUrl` call can just be

```scala
        .uploadFromUrl(
          from = fromUrl,
          to = Path.Directory(s"movements/${movementId.value}").file(s"${movementId.value}-${messageId.value}-$formattedDateTime.xml"),
          owner = "common-transit-convention-traders"
        )
```

The other parameters are defaulted, and we don't need to specify them.

Reply:
@olakunlekuye2  please do the changes as suggested by Dan

Reply:
Sure

Reply:
Done